### DISCUSSION TITLE
How to connect dropzone in fastapi

### DISCUSSION THREAD

**Discussion by CYCNO** [fastapi/fastapi#6162] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
n/a
```

### Description

n/a

### Operating System

Windows

### Operating System Details

window 10 pro

### FastAPI Version

latest

### Python Version

3.9.8

### Additional Context

Is there anyway to connect dropzone in fastapi. I search everywhere but no be able to find it. If someone can help me by posting examples or code it will be helpful 

Thanks

**Comment by JarroVGIT** [comment]
What is dropzone, what have you tried already? Are there specific things you struggle with?

**Comment by CYCNO** [comment]
> 

[dropzone](https://dropzone.dev) is javascript library which help in upload file
i am struggling with saving file in download folder

**Comment by JarroVGIT** [comment, marked as answer]
Well that peeked my interest. Here is a full working example:

```python
import aiofiles
from fastapi import FastAPI, UploadFile
from fastapi.responses import HTMLResponse

app = FastAPI()


@app.get("/")
async def root():
    return {"hello": "world"}


html = """ 
<html>
<head>
    <title>Some Upload Form</title>
    <script src="https://unpkg.com/dropzone@5/dist/min/dropzone.min.js"></script>
    <link rel="stylesheet" href="https://unpkg.com/dropzone@5/dist/min/dropzone.min.css" type="text/css" />
</head>
<body>
    <form action="/uploadfiles" class="dropzone" id="my-great-dropzone">
    </form>
    <script>
        Dropzone.options.myGreatDropzone = { // camelized version of the `id`
            paramName: "uploaded_files", // The name that will be used to transfer the file
            maxFilesize: 2, // MB
        };
    </script>
</body>
</html>
"""


@app.get("/form")
async def present_form():
    return HTMLResponse(html)


@app.post("/uploadfiles")
async def create_upload_files(uploaded_files: list[UploadFile]):
    for file in uploaded_files:
        async with aiofiles.open(f"uploaded_files/{file.filename}", "wb") as out_file:
            content = await file.read()
            await out_file.write(content)
        print(f"File uploaded: {file.filename}")


if __name__ == "__main__":
    import uvicorn

    uvicorn.run(app, host="0.0.0.0", port=8000)
```

The endpoint `/form` will present you with the most basic form there is with Dropzone. It is configured to upload files to the `/uploadfiles` endpoint (which is a post method). 

On the FastAPI side, you will receive a `UploadFile` object, which you can store wherever and name whatever you want. In this example, it is stored in `uploadedfiles/` folder, which is relative from your current working directory (from where Uvicorn is run).

To try this out: run the above code, go to `localhost:8000/form`, drop a file in there and check your Uvicorn logs (it should say "File uploaded: <filename>"). 

Please close this issue if this answers your question.

**Comment by CYCNO** [comment]
> Well that peeked my interest. Here is a full working example:
> 
> ```python
> import aiofiles
> from fastapi import FastAPI, UploadFile
> from fastapi.responses import HTMLResponse
> 
> app = FastAPI()
> 
> 
> @app.get("/")
> async def root():
>     return {"hello": "world"}
> 
> 
> html = """ 
> <html>
> <head>
>     <title>Some Upload Form</title>
>     <script src="https://unpkg.com/dropzone@5/dist/min/dropzone.min.js"></script>
>     <link rel="stylesheet" href="https://unpkg.com/dropzone@5/dist/min/dropzone.min.css" type="text/css" />
> </head>
> <body>
>     <form action="/uploadfiles" class="dropzone" id="my-great-dropzone">
>     </form>
>     <script>
>         Dropzone.options.myGreatDropzone = { // camelized version of the `id`
>             paramName: "uploaded_files", // The name that will be used to transfer the file
>             maxFilesize: 2, // MB
>         };
>     </script>
> </body>
> </html>
> """
> 
> 
> @app.get("/form")
> async def present_form():
>     return HTMLResponse(html)
> 
> 
> @app.post("/uploadfiles")
> async def create_upload_files(uploaded_files: list[UploadFile]):
>     for file in uploaded_files:
>         async with aiofiles.open(f"uploaded_files/{file.filename}", "wb") as out_file:
>             content = await file.read()
>             await out_file.write(content)
>         print(f"File uploaded: {file.filename}")
> 
> 
> if __name__ == "__main__":
>     import uvicorn
> 
>     uvicorn.run(app, host="0.0.0.0", port=8000)
> ```
> 
> The endpoint `/form` will present you with the most basic form there is with Dropzone. It is configured to upload files to the `/uploadfiles` endpoint (which is a post method).
> 
> On the FastAPI side, you will receive a `UploadFile` object, which you can store wherever and name whatever you want. In this example, it is stored in `uploadedfiles/` folder, which is relative from your current working directory (from where Uvicorn is run).
> 
> To try this out: run the above code, go to `localhost:8000/form`, drop a file in there and check your Uvicorn logs (it should say "File uploaded: ").
> 
> Please close this issue if this answers your question.

brooo thank you so much it really work

**Comment by CYCNO** [comment]
hi i have also a question that after uploading file i need to send random quote from python to webpage so how can i do it @JarroVGIT  because i tried it with jinja but it didn't works

**Comment by JarroVGIT** [comment]
You can return a value in FastAPI, something like this:

```
    return {
        "uploaded": [file.filename for file in uploaded_files],
        "random_string": "randomstringhere..",
    }
```

And in your dropzone config, you can catch this on success: 

```
<script>
        Dropzone.options.myGreatDropzone = { 
            paramName: "uploaded_files", 
            maxFilesize: 2, 
            success: function(file, response){
                console.log(response);
            }
        };
    </script>
```

The above sample simply writes it to the console of your browser but you can do whatever with the return value of FastAPI there. But this goes well beyond the scope of this project, so I would suggest to move any further questions related to Dropzone to StackOverflow or to its specific repo.

**Comment by CYCNO** [comment]
> You can return a value in FastAPI, something like this:
> 
> ```
>     return {
>         "uploaded": [file.filename for file in uploaded_files],
>         "random_string": "randomstringhere..",
>     }
> ```
> 
> And in your dropzone config, you can catch this on success:
> 
> ```
> <script>
>         Dropzone.options.myGreatDropzone = { 
>             paramName: "uploaded_files", 
>             maxFilesize: 2, 
>             success: function(file, response){
>                 console.log(response);
>             }
>         };
>     </script>
> ```
> 
> The above sample simply writes it to the console of your browser but you can do whatever with the return value of FastAPI there. But this goes well beyond the scope of this project, so I would suggest to move any further questions related to Dropzone to StackOverflow or to its specific repo.

thanks mate you just save my tons of time

**Comment by tiangolo** [comment]
Wow, you owe @JarroVGIT a coffee haha ☕

Thanks for the discussion, help, and for coming back to close the issue!

**Comment by JarroVGIT** [comment]
I am sorry, this is really far from out of scope of a FastAPI issue. Please post your question on StackOverflow, thanks.

**Comment by CYCNO** [comment]
ok

**Comment by tiangolo** [comment]
Thanks for coming back to close the issue @CYCNO ☕
